lib/srv/mcp/http.go

```diff
@@ -0,0 +1,303 @@
+	return nil
+}
+
+func (s *Server) handleAuthErrHTTP(ctx context.Context, clientConn net.Conn, authErr error) error {
+	return trace.Wrap(s.serveHTTPConn(ctx, clientConn, http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		trace.WriteError(w, authErr)
+	})))
+}
+
+func (s *Server) handleStreamableHTTP(ctx context.Context, sessionCtx *SessionCtx) error {
+	session, err := s.getSessionHandlerWithJWT(ctx, sessionCtx)
+	if err != nil {
+		return trace.Wrap(err, "setting up session handler")
+	}
+
+	transport, err := s.makeStreamableHTTPTransport(session)
+	if err != nil {
+		return trace.Wrap(err, "setting up streamable http transport")
+	}
+
+	session.logger.DebugContext(ctx, "Started handling HTTP request")
+	defer session.logger.DebugContext(ctx, "Completed handling HTTP request")
+
+	delegate := reverseproxy.NewHeaderRewriter()
+	reverseProxy, err := reverseproxy.New(
+		reverseproxy.WithFlushInterval(100*time.Millisecond),
+		reverseproxy.WithRoundTripper(transport),
+		reverseproxy.WithLogger(session.logger),
+		reverseproxy.WithRewriter(appcommon.NewHeaderRewriter(delegate)),
+		reverseproxy.WithResponseModifier(func(resp *http.Response) error {
+			if resp.Request != nil && resp.Request.Method == http.MethodDelete {
+				// Nothing to modify here.
+				return nil
+			}
+			return trace.Wrap(mcputils.ReplaceHTTPResponse(ctx, resp, newHTTPResponseReplacer(session)))
+		}),
+	)
+	if err != nil {
+		return trace.Wrap(err, "creating reverse proxy")
+	}
+
+	return trace.Wrap(s.serveHTTPConn(ctx, sessionCtx.ClientConn, reverseProxy))
+}
+
+func (s *Server) makeStreamableHTTPTransport(session *sessionHandler) (http.RoundTripper, error) {
+	targetURI, err := url.Parse(session.App.GetURI())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	targetURI.Scheme = strings.TrimPrefix(targetURI.Scheme, "mcp+")
+
+	targetTransport, err := s.makeHTTPTransport(session.App)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &streamableHTTPTransport{
+		sessionHandler:  session,
+		targetURI:       targetURI,
+		targetTransport: targetTransport,
+	}, nil
+}
+
+type streamableHTTPTransport struct {
+	*sessionHandler
+	targetURI       *url.URL
+	targetTransport http.RoundTripper
+}
+
+func (t *streamableHTTPTransport) RoundTrip(r *http.Request) (*http.Response, error) {
+	t.setExternalSessionID(r.Header)
+
+	switch r.Method {
+	case http.MethodDelete:
+		return t.handleSessionEndRequest(r)
+	case http.MethodGet:
+		return t.handleListenSSEStreamRequest(r)
+	case http.MethodPost:
+		return t.handleMCPMessage(r)
+
+	default:
+		t.emitInvalidHTTPRequest(t.parentCtx, r)
+		return &http.Response{
+			Request:    r,
+			StatusCode: http.StatusMethodNotAllowed,
+		}, nil
+	}
+}
+
+func (t *streamableHTTPTransport) setExternalSessionID(header http.Header) {
+	if id := header.Get(mcpSessionIDHeader); id != "" {
+		t.mcpSessionID.Store(&id)
+	}
+}
+
+func (t *streamableHTTPTransport) rewriteRequest(r *http.Request) *http.Request {
```

Comment:
Should we consistently enforce the request to be `POST` or `GET`? We could raise errors in case any other method is requested.